# Book Startup

1. Introduction
* Overview of the project
* Goals and objectives

2. Data / Server Loading 
* Import necessary packets
* Connect to the server

3. Queries


4. Conclusions and Recommendations


## Introduction

The coronavirus surprised the entire world, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and shopping centers, but more people stayed home reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

We will work with a database of one of the services competing in this market. It contains data about books, publishers, authors, customer ratings, and book reviews. This information will be used to generate a value proposition for a new product.

### Data / Server Loading 

In [2]:
import sqlalchemy
print("sqlalchemy is installed and working!")

sqlalchemy is installed and working!


In [3]:
# importar librerías
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Queries

### Analyze the tables

In [4]:
# Check the first five rows from books table

query_books = """
SELECT * FROM books
LIMIT 5;
"""

result_books = pd.read_sql(query_books, con=engine)
print("First 5 rows from books table:")
display(result_books)

First 5 rows from books table:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# Check the first five rows from authors table

query_authors = """
SELECT * FROM authors
LIMIT 5;
"""

result_authors = pd.read_sql(query_authors, con=engine)
print("First 5 rows from authors table:")
display(result_authors)

First 5 rows from authors table:


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# Check the first five rows from publishers table

query_publishers = """
SELECT * FROM publishers
LIMIT 5;
"""
result_publishers = pd.read_sql(query_publishers, con=engine)
print("First 5 rows from publishers table:")
display(result_publishers)

First 5 rows from publishers table:


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [7]:
#Check the first five rows from ratings table"
query_ratings = """
SELECT * FROM ratings
LIMIT 5;
"""
result_ratings = pd.read_sql(query_ratings, con=engine)
print("First 5 rows from ratings table:")
display(result_ratings)

First 5 rows from ratings table:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
query_reviews = """
SELECT * FROM reviews
LIMIT 5;
"""
result_reviews = pd.read_sql(query_reviews, con=engine)
print("First 5 rows from reviews table:")
display(result_reviews)

First 5 rows from reviews table:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Queries

In [9]:
# Query 1: Find the number of books published after January 1, 2000
query1 = """
SELECT COUNT(*) AS books_published_after_2000
FROM books
WHERE publication_date > '2000-01-01';
"""
result1 = pd.read_sql(query1, con=engine)
display(result1)

books_published_after_2000
0                         819

 A total of 819 books in the database were published after January 1, 2000. This indicates that the database has a significant collection of relatively recent publications, suggesting its relevance for users interested in contemporary literature.

 Let's compare to the total number of books in the database to determine the proportion of modern books versus older publications:

In [10]:
# Query: Total number of books in the database
query_total_books = """
SELECT COUNT(*) AS total_books
FROM books;
"""
result_total_books = pd.read_sql(query_total_books, con=engine)
total_books = result_total_books['total_books'][0]

# Query: Number of books published after January 1, 2000
query_books_after_2000 = """
SELECT COUNT(*) AS books_published_after_2000
FROM books
WHERE publication_date > '2000-01-01';
"""
result_books_after_2000 = pd.read_sql(query_books_after_2000, con=engine)
books_published_after_2000 = result_books_after_2000['books_published_after_2000'][0]

# Calculate the proportion of modern books
proportion_modern_books = books_published_after_2000 / total_books * 100

# Display the results
print(f"Total number of books: {total_books}")
print(f"Number of books published after January 1, 2000: {books_published_after_2000}")
print(f"Proportion of modern books: {proportion_modern_books:.2f}%")

Total number of books: 1000
Number of books published after January 1, 2000: 819
Proportion of modern books: 81.90%


* Relevance to Contemporary Readers: The high proportion of modern books indicates that the database is highly relevant to readers interested in recent publications.

* Current Trends: The collection aligns well with current literary trends, potentially attracting a younger audience or readers looking for the latest titles and contemporary authors.

* Marketing Focus: Marketing efforts can highlight the modernity of the collection. Emphasizing the availability of recent books in promotional materials could attract more users.

* Only 18.10% of the collection consists of books published before 2000, which means there is a relatively smaller selection of classic or older literature. While the modern focus is beneficial, ensuring a balanced collection with classic literature can cater to a broader audience, including those interested in older, timeless works. Consider acquiring more classic books to diversify the collection. This will attract readers interested in literary history, academia, or those who enjoy classic genres


 Let's implement a function for **Query** 2 to find the number of user reviews and the average rating for each book. This query can be useful in multiple contexts, such as generating reports for different subsets of books or for books published in different time periods.

In [11]:
# Query 2: Find the number of user reviews and the average rating for each book

def get_book_reviews_and_ratings(engine, start_date=None, end_date=None):
    """
    Retrieves the number of user reviews and average rating for each book.
    
    Parameters:
        engine (sqlalchemy.engine.Engine): SQLAlchemy engine object for the database connection.
        start_date (str): Optional. The start date for filtering books (format: 'YYYY-MM-DD').
        end_date (str): Optional. The end date for filtering books (format: 'YYYY-MM-DD').
        
    Returns:
        pd.DataFrame: DataFrame containing book_id, title, review_count, and average_rating.
    """
    # Base query
    query = """
    SELECT b.book_id, b.title, 
           COUNT(rw.review_id) AS review_count, 
           AVG(rt.rating) AS average_rating
    FROM books b
    LEFT JOIN reviews rw ON b.book_id = rw.book_id
    LEFT JOIN ratings rt ON b.book_id = rt.book_id
    WHERE 1=1
    """
    
    # Add date filters if provided
    if start_date:
        query += f" AND b.publication_date >= '{start_date}'"
    if end_date:
        query += f" AND b.publication_date <= '{end_date}'"
    
    # Group by and order by clauses
    query += """
    GROUP BY b.book_id, b.title
    ORDER BY b.book_id;
    """
    
    # Execute the query
    result = pd.read_sql(query, con=engine)
    return result


engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

# Retrieve reviews and ratings for all books
all_books_reviews_and_ratings = get_book_reviews_and_ratings(engine)
display(all_books_reviews_and_ratings)

book_id                                              title  review_count  \
0          1                                       'Salem's Lot             6   
1          2                 1 000 Places to See Before You Die             2   
2          3  13 Little Blue Envelopes (Little Blue Envelope...             9   
3          4  1491: New Revelations of the Americas Before C...             4   
4          5                                               1776            24   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)             9   
996      997                        Xenocide (Ender's Saga  #3)            15   
997      998                                    Year of Wonders            20   
998      999                        You Suck (A Love Story  #2)             4   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...            24   

     average_rating  
0          3.666667  
1          2.500000  
2          4.666667  
3          4.500000  
4          4.000000  
..              ...  
995        3.666667  
996        3.400000  
997        3.200000  
998        4.500000  
999        3.833333  

[1000 rows x 4 columns]

From the snapshot of **this** ouptput we can see:

**Highly Reviewed Books:**

* "1776" and "Zen and the Art of Motorcycle Maintenance: An Inquiry into Values" both have 24 reviews, indicating high engagement from readers.
* "Year of Wonders" and "Xenocide (Ender's Saga #3)" also have a significant number of reviews, with 20 and 15 respectively.

**Highly Rated Books:**

* "13 Little Blue Envelopes (Little Blue Envelope Series)" has an average rating of 4.67, suggesting it is well-received by readers.
* "1491: New Revelations of the Americas Before Columbus" and "You Suck (A Love Story #2)" both have high average ratings of 4.50.

**Low Average Ratings:**

* "1 000 Places to See Before You Die" has a relatively low average rating of 2.50, indicating mixed or poor reviews.
* "Year of Wonders" has a lower average rating of 3.20 despite having a high number of reviews.

**Books with Moderate Engagement and Ratings:**

* "'Salem's Lot" has a moderate number of reviews (6) with a decent average rating (3.67).
* "Wyrd Sisters (Discworld #6; Witches #2)" also fits this category with 9 reviews and a 3.67 average rating.

These insights suggest that while some books have high engagement and favorable reviews, others might be less well-received or have polarized opinions among readers.


In [12]:
# Query 3: Identify the publisher that has published the greatest number of books with more than 50 pages
query3 = """
SELECT p.publisher, COUNT(b.book_id) AS book_count
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY book_count DESC
LIMIT 1;
"""
result3 = pd.read_sql(query3, con=engine)
display(result3)

publisher  book_count
0  Penguin Books          42

Penguin Books has published 42 books with over 50 pages, making it the leading publisher in this category. It significantly contributes to the database's collection. This dominance implies that Penguin Books is a reliable source of quality publications. Since Penguin has a reputation for its diverse and high-quality literature, its presence in the database can increase user trust and satisfaction, as readers often associate the brand with reputable content.

In [13]:
# Query 4: Identify the author who has the highest average book rating for books with at least 50 ratings
query4 = """
SELECT a.author, 
    AVG(rt.rating) AS average_rating,
    COUNT(rt.rating_id) AS num_ratings
FROM books b
JOIN authors a  ON b.author_id = a.author_id
JOIN ratings rt ON b.book_id = rt.book_id
GROUP by a.author
HAVING COUNT(rt.rating_id) >=50
ORDER BY average_rating DESC
LIMIT 1;
"""

result4 = pd.read_sql(query4, con=engine)
display(result4)


author  average_rating  num_ratings
0  Diana Gabaldon             4.3           50

Diana Gabaldon has the highest average rating for her books, with at least 50 ratings. This indicates that Gabaldon's books are well received by readers. Also the number of reviews shows that her readership is higly engaged with her work. Her high rating and engagement make her a prime candidate for focused marketing efforts.

In [14]:
# Query 5: Find the average number of text reviews among users who rated more than 50 books
query5 = """
WITH user_ratings AS (
    SELECT username, COUNT(rating_id) AS rating_count
    FROM ratings
    GROUP BY username
    HAVING COUNT(rating_id) > 50
)
SELECT AVG(user_reviews.review_count) AS average_review_count
FROM (
    SELECT rw.username, COUNT(rw.review_id) AS review_count
    FROM reviews rw
    JOIN user_ratings ur ON rw.username = ur.username
    GROUP BY rw.username
) user_reviews;
"""
result5 = pd.read_sql(query5, con=engine)
display(result5)

average_review_count
0             24.333333

On average, users who have rated more than 50 books have written approximately 24.33 text reviews. Users who rate many books also actively contribute written reviews. This indicates a highly engaged user base that rates books and takes the time to write detailed reviews. The high average number of reviews suggests that these active users provide substantial feedback, which can be invaluable for other readers and understanding user preferences. The significant average number of reviews among active raters presents an opportunity to encourage more users to write reviews.